internal/manager/run.go

```diff
@@ -52,14 +51,54 @@ import (
 // Controller Manager - Setup & Run
 // -----------------------------------------------------------------------------
 
-// Run starts the controller manager and blocks until it exits.
-func Run(
+type Manager struct {
+	m                    manager.Manager
+	synchronizer         *dataplane.Synchronizer
+	stopAnonymousReports func()
+}
+
+// Run starts the Kong Ingress Controller. It blocks until the context is cancelled.
+// It should be called only once per Manager instance.
+func (m *Manager) Run(ctx context.Context) error {
+	return m.m.Start(ctx)
+}
+
+// IsReady checks if the controller manager is ready to manage resources.
+// It's only valid to call this method after the controller manager has been started
+// with method Run(ctx).
+func (m *Manager) IsReady() error {
+	select {
+	// If we're elected as leader then report readiness based on the readiness
+	// of dataplane synchronizer.
+	case <-m.m.Elected():
+		if !m.synchronizer.IsReady() {
+			return errors.New("synchronizer not yet configured")
+		}
+	// If we're not the leader then just report as ready.
+	default:
+	}
+	return nil
+}
+
+// StopAnonymousReports stops the telemetry reporting. It's caller responsibility to call it when
+// the manager is no longer needed. It's safe to call it multiple times, when the telemetry
+// was not configured, or when the manager is already stopped (it will be no-op).
+// It makes sense to call it only after Run(ctx) method.
+func (m *Manager) StopAnonymousReports() {
```

Comment:
I believe we shouldn't expose it even here (in the `internalmanager.Manager`) as that should be transparent to the caller how the manager tears down after the context gets canceled, and no additional calls should be expected. I understand we need to somehow tear down reporting goroutines as they are now started in the constructor, and we cannot rely on a `defer`. How about hiding it in the `Manager.Run` (set up a defer fn calling `stopAnonymousReports`) instead of exposing a public method?

As part of https://github.com/Kong/kubernetes-ingress-controller/issues/7041 we could extend the telemetry manager to accept a context in its `Start` method so it behaves as other goroutines so we don't need to tear it down explicitly.

Reply:
You're right I missed it that I can do it in `Run(...)` from this manager, fixing